packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts

```diff
@@ -220,6 +232,14 @@ abstract class PortfolioBase extends cdk.Resource implements IPortfolio {
     AssociationManager.deployWithStackSets(this, product, options);
   }
 
+  public giveAssetBucketsReadToSharedAccounts() {
+    for (const bucket of this.assetBuckets) {
+      for (const accountId of this.sharedAccounts) {
+        bucket.grantRead(new iam.AccountPrincipal(accountId));
+      }
```

Comment:
I'm not exactly sure what the purpose of this method is, but it might be easier/better to use a `CompositePrincipal` instead for the policy.

Reply:
The goal is for the portfolio to collect a set of S3 Asset buckets and a list of shared accounts. Given that the portfolio does not have an `onSynthesize` method, we created an aspect that will take each bucket and grant read to all accounts.  Essentially, in the scenario where the portfolio contains three asset buckets and two accounts, we would want the portfolio to iterate through each bucket and grant read permissions to the two accounts. If `CompositePrincipal` can eliminate that second for loop that would be great, will definitely look into it

Reply:
How does this look? Hopefully it's more readable.
```
    for (const bucket of this.assetBuckets) {
      const compositeAccountPrincipals = new iam.CompositePrincipal(...this.sharedAccounts.map(
        (accountId) => new iam.AccountPrincipal(accountId))
      );
      bucket.grantRead(compositeAccountPrincipals);
    }
```

Reply:
I have a question here. We have the following scenario: our stack creates multiple porfolios containing one product. Each portfolio is shared with other account but sometimes the products have exaclty the same asset (reusable lambda in our case). Would that mean that for each portfolio there will be a separate bucket created?

Reply:
Great question, this was actually the initial design goal - one product stack per portfolio. I considered working with the cdk bootstrap bucket but the idea of giving cross account access to a bucket containing all of the user's assets did not seem to adhere to best security practices. 

Ideally, we would like to reduce the share scope as much as possible. Portfolios seemed like a great way to accomplish this until we realized the timeline of events would not allow for us to create an asset bucket in the portfolio and pass it down to the product stack? Why? Well, the product stack needs the s3 bucket containing the asset buckets at time of creation so that the product stack synthesizer is able to pass back the expected location of the asset in its ` public addFileAsset(_asset: cdk.FileAssetSource): cdk.FileAssetLocation` method. 
```
Portfolio constructor
ProductStack constructor
ProductStackSynthesizer::bind()
ProductStackSynthesizer::addFileAsset()
ProductStack::addFileAssetToParentSynthesizer() 
	This is where we need the product stack to know the name of the bucket
CloudFormationProductStackTemplate constructor
CloudFormationProduct constructor
CloudFormationProductStackTemplate::bind()
Portfolio::addProduct()
Portfolio::shareWithAccount()
ProductStackSynthesizer::synthesize()
```

We could technically make the product stack asset bucket external facing and have the user create the bucket in their application stack. This would require allowing product stack props so that we can take the asset bucket as a prop and proceed from there. This approach was less favored by our team given that it would force users to interact with their product stack differently from the way they already do **(if they want to use assets)**. Here is what this user experience would look like:

```
export class HelloServerlessStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create development bucket
    const devBucket = new ProductStackAssetBucket(this, "ProductStackAssetBucket");

    // Create a portfolio
    const portfolio = new sc.Portfolio(this, 'DevToolsPortfolio', {
      displayName: 'DevTools', 
      providerName: 'IT', 
      assetBucket: devBucket, 
    });

    // Create a Lambda product from a Product Stack
    const product = new sc.CloudFormationProduct(this, 'HelloServerless', {
      productName: "HelloServerless",
      owner: "IT",
      productVersions: [
        {
           cloudFormationTemplate: sc.CloudFormationTemplate.fromProductStack(
            new HelloServerlessProduct(this, 'HelloServerlessProduct', {assetBucket: devBucket})),
          productVersionName: "v1",
          description: "A simple REST API backed by a hello Lambda function"
        },
      ],
    });
```

Reply:
Actually I like this approach. Would it be possible to combine Your's approach with this You introduced here, but the ```assetBucket``` property would be optional? This would be a good solution for the use case we are facing with my team.
The other one would simply build over 100 buckets (soft quota).
I imagine we could create this AssetsBucket in a separate stack that would be deployed right before the portfolio stack and this would be in our case not a security concern. Actually our current apprach uses the cdk assets bucket, so I actually understand Your's considerations about security, because we also had this discussions already.

Reply:
With even more trickery, you could also return a `Lazy.string()` as bucketName in the `Synthesizer`, and copy in the the real bucket name when the `ProductStack` is being synthesized in the context of a Portfolio.

(Btw. `Lazy.string()` will lead to every `ProductStack` instance only being usable in the context of a single `Portfolio`. Not sure how desirable or how bad that is, but you might want to add validation to make sure users don't get it wrong. Alternatively, `Lazy.uncachedString()` might help, but needs a test to confirm that it works if that's what you want to support)

---------

> I considered working with the cdk bootstrap bucket but the idea of giving cross account access to a bucket containing all of the user's assets did not seem to adhere to best security practices.

Good call 😉. I like the approach you're taking here.

> Portfolios seemed like a great way to accomplish this until we realized the timeline of events would not allow for us to create an asset bucket in the portfolio and pass it down to the product stack

In software, things are rarely as cut-and-dried as "it's not possible". Most things are possible (given an appropriate amount of work). You've identified the ideal customer experience already; now is the time to fight to see if you can achieve it.

Reply:
> Actually I like this approach. Would it be possible to combine Your's approach with this You introduced here, but the `assetBucket` property would be optional? This would be a good solution for the use case we are facing with my team. The other one would simply build over 100 buckets (soft quota). I imagine we could create this AssetsBucket in a separate stack that would be deployed right before the portfolio stack and this would be in our case not a security concern. Actually our current apprach uses the cdk assets bucket, so I actually understand Your's considerations about security, because we also had this discussions already.

Great, thanks for the feedback. We changed our implementation to support the user passing in an optional product stack asset bucket. The user will also have the ability to pass in a name for the assets bucket (otherwise, we'll generate the name for you 🙂).